lib/data-access/bookmarks.ts

```
@@ -0,0 +1,111 @@
+
+/**
+ * Retrieves bookmark data using a hierarchical strategy: in-memory cache, S3 storage, and external API as fallback.
+ *
+ * If {@link skipExternalFetch} is true, only cache and S3 are used; otherwise, the function attempts to fetch from the external source and updates both S3 and cache on success. Returns an empty array if no bookmarks are found.
+ *
+ * @param skipExternalFetch - If true, bypasses the external API and relies solely on cache or S3. Defaults to false.
+ * @returns A promise resolving to an array of UnifiedBookmark objects, or an empty array if none are available.
+ */
+export async function getBookmarks(skipExternalFetch = false): Promise<UnifiedBookmark[]> {
+  const cached = ServerCacheInstance.getBookmarks();
+  if (cached && cached.bookmarks && cached.bookmarks.length > 0 && skipExternalFetch) {
```

Biome (1.9.4)

[error] Change to an optional chain.

Unsafe fix: Change to an optional chain.


(lint/complexity/useOptionalChain)